src/pickle-jar-univ3.sol

```diff
@@ -0,0 +1,255 @@
+        controller = _controller;
+        paused = false;
+        earnAfterDeposit = true;
+    }
+
+    function totalLiquidity() public view returns (uint256) {
+        return liquidityOfThis().add(IControllerV2(controller).liquidityOf(address(pool)));
+    }
+
+    function liquidityOfThis() public view returns (uint256) {
+        uint256 _balance0 = token0.balanceOf(address(this));
+        uint256 _balance1 = token1.balanceOf(address(this));
+        return uint256(pool.liquidityForAmounts(_balance0, _balance1, getLowerTick(), getUpperTick()));
+    }
+
+    function getUpperTick() public view returns (int24) {
+        return IControllerV2(controller).getUpperTick(address(pool));
+    }
+
+    function getLowerTick() public view returns (int24) {
+        return IControllerV2(controller).getLowerTick(address(pool));
+    }
+
+    function setGovernance(address _governance) public {
+        require(msg.sender == governance, "!governance");
+        governance = _governance;
+    }
+
+    function setTimelock(address _timelock) public {
+        require(msg.sender == timelock, "!timelock");
+        timelock = _timelock;
+    }
+
+    function setController(address _controller) public {
+        require(msg.sender == timelock, "!timelock");
+        controller = _controller;
+    }
+
+    event JarPaused(uint256 block, uint256 timestamp);
+
+    function setPaused(bool _paused) external {
+        require(msg.sender == governance, "!governance");
+        paused = _paused;
+        emit JarPaused(block.number, block.timestamp);
+    }
+
+    function setEarnAfterDeposit(bool _earnAfterDeposit) external {
+        require(msg.sender == governance, "!governance");
+        earnAfterDeposit = _earnAfterDeposit;
+    }
+
+    function earn() public {
+        require(liquidityOfThis() > 0, "no liquidity here");
+
+        uint256 balance0 = token0.balanceOf(address(this));
+        uint256 balance1 = token1.balanceOf(address(this));
+
+        token0.safeTransfer(controller, balance0);
+        token1.safeTransfer(controller, balance1);
+
+        IControllerV2(controller).earn(address(pool), balance0, balance1);
+    }
+
+    function deposit(uint256 token0Amount, uint256 token1Amount) external payable nonReentrant whenNotPaused {
+        bool isEthToken0;
+        bool isEthToken1;
+        uint256 _eth = address(this).balance;
+        if (_eth > 0) {
+            WETH(weth).deposit{value: _eth}();
+            if(address(token0) == weth){
+              token0Amount = _eth;
+              isEthToken0 = true;
+            }
+            else if(address(token1) == weth) {
+              token1Amount = _eth;
+              isEthToken1 = true;
+            }
+        }
+
+        if (token0Amount > 0 && !isEthToken0) token0.safeTransferFrom(msg.sender, address(this), token0Amount);
+        if (token1Amount > 0 && !isEthToken1) token1.safeTransferFrom(msg.sender, address(this), token1Amount);
+
+        balanceProportion();
```

Comment:
I think that the `balancePropotion()` call would be more appropriate in the strategy's `deposit` function, immediately preceding the `increaseLiquidity` call. This serves to 1) isolate the rebalancing logic in the strategy (at the point of adding liquidity to avoid intervening state mutation), and 2) potentially avoids an expensive call if `earnAfterDeposit` is ever set to false.

Reply:
I agree, but would like to discuss this with you.